task/orchestrator.go

```diff
@@ -18,35 +18,54 @@ import (
 )
 
 type Orchestrator struct {
-	config   Config
-	configIn io.Reader
+	config                 Config
+	configIn               io.Reader
+	terminationGracePeriod time.Duration
 
 	ready      atomic.Bool
 	agentPid   atomic.Int64
 	cancelTask context.CancelFunc
 	cancelMu   sync.RWMutex
 }
 
-func NewOrchestrator(configIn io.Reader) *Orchestrator {
+func NewOrchestrator(configIn io.Reader, terminationGracePeriod time.Duration) *Orchestrator {
 	return &Orchestrator{
-		config:     Config{},
-		configIn:   configIn,
-		cancelTask: func() {},
+		config:                 Config{},
+		configIn:               configIn,
+		cancelTask:             func() {},
+		terminationGracePeriod: terminationGracePeriod,
 	}
 }
 
-func (o *Orchestrator) Run(ctx context.Context) error {
-	ctx = o.taskContext(ctx)
+func (o *Orchestrator) Run(parentCtx context.Context) (err error) {
```

Comment:
This allows us to handle shutdown a bit more gracefully. When the `system` receives a termination signal, it'll cancel the parent context. With this PR, we catch that here and will wait for the configured grace period before fully shutting down. This gives the chance for the task to complete if it's within the Pod's `terminationGracePeriodSeconds`. Otherwise, the task will be interrupted.